[Elastic Agent] Add support for variable replacement from providers

blakerouse opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

Adds support for replacing `{{local_dynamic.key|local_dynamic.other|'fallback'}}` inside of the Elastic Agent `inputs` configuration.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Finishes the core support for #20781 

More code is needed to help debugging effort.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

Running the Elastic Agent in local mode, use the following configuration:

```
outputs:
  default:
    type: elasticsearch
    hosts: [127.0.0.1:9200]
    username: elastic
    password: changeme

providers:
  local_dynamic:
    items:
      - vars:
          key: value1
        processors:
          - add_fields:
              fields:
                custom: match1
              target: dynamic
      - vars:
          key: value2
        processors:
          - add_fields:
              fields:
                custom: match2
              target: dynamic
      - vars:
          key: value3
        processors:
          - add_fields:
              fields:
                custom: match3
              target: dynamic

inputs:
  - type: logfile
    enabled: true
    streams:
      - paths:
          - /var/log/{{local_dynamic.key}}
```

Then run the following to inspect the generated configuration:

```
$ ./elastic_agent inspect output -o default
[default] filebeat:
filebeat:
  inputs:
  - index: logs-generic-default
    paths:
    - /var/log/value1
    processors:
    - add_fields:
        fields:
          custom: match1
        target: dynamic
    - add_fields:
        fields:
          dataset: generic
          namespace: default
          type: logs
        target: data_stream
    - add_fields:
        fields:
          dataset: generic
        target: event
    type: log
  - index: logs-generic-default
    paths:
    - /var/log/value2
    processors:
    - add_fields:
        fields:
          custom: match2
        target: dynamic
    - add_fields:
        fields:
          dataset: generic
          namespace: default
          type: logs
        target: data_stream
    - add_fields:
        fields:
          dataset: generic
        target: event
    type: log
  - index: logs-generic-default
    paths:
    - /var/log/value3
    processors:
    - add_fields:
        fields:
          custom: match3
        target: dynamic
    - add_fields:
        fields:
          dataset: generic
          namespace: default
          type: logs
        target: data_stream
    - add_fields:
        fields:
          dataset: generic
        target: event
    type: log
output:
  elasticsearch:
    hosts:
    - 127.0.0.1:9200
    password: changeme
    username: elastic

---
[default] FLEET_MONITORING:
output:
  elasticsearch:
    hosts:
    - 127.0.0.1:9200
    password: changeme
    type: elasticsearch
    username: elastic
programs:
- filebeat

---
```

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Relates #20781

elasticmachine commented:
Pinging @elastic/ingest-management (Team:Ingest Management)

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20839/detail/PR-20839/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20839/detail/PR-20839/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20839/detail/PR-20839/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20839/detail/PR-20839/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #20839 updated]`

* Reason: `Aborted from #3`

* Start Time: 2020-08-27T19:57:13.105+0000

* Duration: 10 min 3 sec


* Commit: 96bc36c6770a1a5ad7d66f661d8d1a7c5c9471c8




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->




<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-08-27T20:06:29.631Z] Failed in branch Elastic Agent x-pack
[2020-08-27T20:06:29.645Z] Failed in branch Elastic Agent x-pack Windows
[2020-08-27T20:06:29.656Z] Failed in branch Elastic Agent Mac OS X
[2020-08-27T20:06:29.668Z] Failed in branch Filebeat oss
[2020-08-27T20:06:29.681Z] Failed in branch Filebeat x-pack
[2020-08-27T20:06:29.694Z] Failed in branch Filebeat Mac OS X
[2020-08-27T20:06:29.704Z] Failed in branch Filebeat x-pack Mac OS X
[2020-08-27T20:06:29.716Z] Failed in branch Filebeat Windows
[2020-08-27T20:06:29.733Z] Failed in branch Filebeat x-pack Windows
[2020-08-27T20:06:29.748Z] Failed in branch Heartbeat oss
[2020-08-27T20:06:29.759Z] Failed in branch Heartbeat Mac OS X
[2020-08-27T20:06:29.769Z] Failed in branch Heartbeat Windows
[2020-08-27T20:06:29.778Z] Failed in branch Auditbeat oss Linux
[2020-08-27T20:06:29.788Z] Failed in branch Auditbeat crosscompile
[2020-08-27T20:06:29.797Z] Failed in branch Auditbeat oss Mac OS X
[2020-08-27T20:06:29.808Z] Failed in branch Auditbeat oss Windows
[2020-08-27T20:06:29.818Z] Failed in branch Auditbeat x-pack
[2020-08-27T20:06:29.830Z] Failed in branch Auditbeat x-pack Mac OS X
[2020-08-27T20:06:29.840Z] Failed in branch Auditbeat x-pack Windows
[2020-08-27T20:06:29.851Z] Failed in branch Libbeat x-pack
[2020-08-27T20:06:29.860Z] Failed in branch Metricbeat OSS Unit tests
[2020-08-27T20:06:29.869Z] Failed in branch Metricbeat OSS Go Integration tests
[2020-08-27T20:06:29.878Z] Failed in branch Metricbeat OSS Python Integration tests
[2020-08-27T20:06:29.887Z] Failed in branch Metricbeat crosscompile
[2020-08-27T20:06:29.897Z] Failed in branch Metricbeat Mac OS X
[2020-08-27T20:06:29.908Z] Failed in branch Metricbeat x-pack Mac OS X
[2020-08-27T20:06:29.917Z] Failed in branch Metricbeat Windows
[2020-08-27T20:06:29.926Z] Failed in branch Metricbeat x-pack Windows
[2020-08-27T20:06:29.935Z] Failed in branch Packetbeat Linux
[2020-08-27T20:06:29.944Z] Failed in branch Packetbeat Mac OS X
[2020-08-27T20:06:29.954Z] Failed in branch Packetbeat Windows
[2020-08-27T20:06:29.967Z] Failed in branch dockerlogbeat
[2020-08-27T20:06:29.977Z] Failed in branch Winlogbeat oss
[2020-08-27T20:06:29.988Z] Failed in branch Winlogbeat Windows
[2020-08-27T20:06:30.000Z] Failed in branch Winlogbeat Windows x-pack
[2020-08-27T20:06:30.010Z] Failed in branch Functionbeat x-pack
[2020-08-27T20:06:30.023Z] Failed in branch Functionbeat Mac OS X x-pack
[2020-08-27T20:06:30.037Z] Failed in branch Functionbeat Windows
[2020-08-27T20:06:30.050Z] Failed in branch Journalbeat
[2020-08-27T20:06:30.060Z] Failed in branch Generators Metricbeat Mac OS X
[2020-08-27T20:06:30.070Z] Failed in branch Generators Beat Mac OS X
[2020-08-27T20:06:30.081Z] Failed in branch Kubernetes
[2020-08-27T20:06:30.408Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-08-27T20:06:30.419Z] Stage "Metricbeat x-pack" skipped due to earlier failure(s)
[2020-08-27T20:06:30.431Z] Stage "Generators" skipped due to earlier failure(s)
[2020-08-27T20:06:31.371Z] Failed in branch Metricbeat x-pack
[2020-08-27T20:06:31.381Z] Failed in branch Generators
[2020-08-27T20:06:31.434Z] Stage "Libbeat" skipped due to earlier failure(s)
[2020-08-27T20:06:31.654Z] Failed in branch Libbeat
[2020-08-27T20:06:37.256Z] [INFO] unstashV2: JOB_GCS_BUCKET is set. bucket param got precedency instead.
[2020-08-27T20:06:37.300Z] [INFO] unstashV2: JOB_GCS_CREDENTIALS is set. credentialsId param got precedency instead.
[2020-08-27T20:06:37.360Z] [Google Cloud Storage Plugin] Found 1 files to download from pattern: gs://beats-ci-temp/Beats/beats/PR-20839-1/source/source.tgz
[2020-08-27T20:06:37.378Z] [Google Cloud Storage Plugin] Downloading: Beats/beats/PR-20839-1/source/source.tgz to local path: /var/lib/jenkins/workspace/Beats_beats_PR-20839/source.tgz
[2020-08-27T20:06:44.487Z] + tar --version
[2020-08-27T20:06:44.800Z] + tar -xpf source.tgz
[2020-08-27T20:06:55.216Z] + rm source.tgz
[2020-08-27T20:06:55.244Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-20839/src/github.com/elastic/beats
[2020-08-27T20:06:55.307Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-20839/src/github.com/elastic/beats/Lint
[2020-08-27T20:06:55.871Z] + cat
[2020-08-27T20:06:55.871Z] + /usr/local/bin/runbld ./runbld-script --job-name elastic+beats+pull-request
[2020-08-27T20:06:55.871Z] Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
[2020-08-27T20:07:02.604Z] runbld>>> runbld started
[2020-08-27T20:07:02.604Z] runbld>>> 1.6.12/f45d832f2ba0aa2722ab4ec1fda8ad140f027f8b
[2020-08-27T20:07:03.543Z] runbld>>> The following profiles matched the job 'elastic+beats+pull-request' in order of occurrence in the config (last value wins).
[2020-08-27T20:07:03.543Z] runbld>>> Matches in the system config:
[2020-08-27T20:07:03.543Z] runbld>>> - Matched ^elastic\+beats
[2020-08-27T20:07:03.543Z] runbld>>> - Matched ^elastic\+beats\+pull-request
[2020-08-27T20:07:04.920Z] runbld>>> Debug logging enabled.
[2020-08-27T20:07:04.920Z] runbld>>> Storing result
[2020-08-27T20:07:04.920Z] runbld>>> Store result: created {:total 2, :successful 2, :failed 0} 1
[2020-08-27T20:07:04.920Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1597739501209/t/20200827200704-9D71F623
[2020-08-27T20:07:04.920Z] runbld>>> Adding system facts.
[2020-08-27T20:07:05.862Z] runbld>>> Adding vcs info for the latest commit:  79135c6ddc423c7332bdc8d0bbc7e327b3a40c0d
[2020-08-27T20:07:05.862Z] runbld>>> >>>>>>>>>>>> SCRIPT EXECUTION BEGIN >>>>>>>>>>>>
[2020-08-27T20:07:05.862Z] runbld>>> Adding /usr/lib/jvm/java-8-openjdk-amd64/bin to the path.
[2020-08-27T20:07:05.863Z] + echo 'Processing JUnit reports with runbld...'
[2020-08-27T20:07:05.863Z] Processing JUnit reports with runbld...
[2020-08-27T20:07:06.431Z] runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
[2020-08-27T20:07:06.431Z] runbld>>> DURATION: 24ms
[2020-08-27T20:07:06.431Z] runbld>>> STDOUT: 40 bytes
[2020-08-27T20:07:06.431Z] runbld>>> STDERR: 49 bytes
[2020-08-27T20:07:06.431Z] runbld>>> WRAPPED PROCESS: SUCCESS (0)
[2020-08-27T20:07:06.431Z] runbld>>> Searching for build metadata in /var/lib/jenkins/workspace/Beats_beats_PR-20839
[2020-08-27T20:07:07.366Z] runbld>>> Storing build metadata: 
[2020-08-27T20:07:07.366Z] runbld>>> Adding test report.
[2020-08-27T20:07:07.366Z] runbld>>> Searching for junit test output files with the pattern: TEST-.*\.xml$ in: /var/lib/jenkins/workspace/Beats_beats_PR-20839/src/github.com/elastic/beats
[2020-08-27T20:07:07.936Z] runbld>>> Found 0 test output files
[2020-08-27T20:07:07.936Z] runbld>>> Test output logs contained: Errors: 0 Failures: 0 Tests: 0 Skipped: 0
[2020-08-27T20:07:07.936Z] runbld>>> Storing result
[2020-08-27T20:07:08.196Z] runbld>>> Store result: updated {:total 2, :successful 2, :failed 0} 2
[2020-08-27T20:07:08.196Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1597739501209/t/20200827200704-9D71F623
[2020-08-27T20:07:08.454Z] runbld>>> Email notification disabled by environment variable.
[2020-08-27T20:07:08.454Z] runbld>>> Slack notification disabled by environment variable.
[2020-08-27T20:07:14.124Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-20839
[2020-08-27T20:07:14.337Z] [INFO] getVaultSecret: Getting secrets
[2020-08-27T20:07:14.433Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-08-27T20:07:15.960Z] + chmod 755 generate-build-data.sh
[2020-08-27T20:07:15.960Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20839/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20839/runs/1 ABORTED 602580
[2020-08-27T20:07:15.960Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20839/runs/1/steps/?limit=10000 -o steps-info.json
[2020-08-27T20:07:16.210Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20839/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ruflin commented on x-pack/elastic-agent/pkg/agent/transpiler/ast.go:
```diff
@@ -193,14 +216,14 @@ func NewList(nodes []Node) *List {
 
 func (l *List) String() string {
 	var sb strings.Builder
+	sb.WriteString("[")
```
Was this a "bug" before?

blakerouse replied:
Yes it was. We only `ast.String()` for testing and development, so it doesn't have an effect on operations or even log output.

ruflin commented on x-pack/elastic-agent/pkg/agent/transpiler/ast_test.go:
```diff
@@ -704,6 +706,123 @@ func TestSelector(t *testing.T) {
 	}
 }
 
+func TestAST_Apply(t *testing.T) {
```
Some tests with golden files could be useful here to make it easier to follow.

blakerouse replied:
Yes that would have made it easier to read.

ruflin commented on x-pack/elastic-agent/pkg/composable/vars.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+// splitPipes split value at |, ignoring | in quotes.
+func splitPipes(input string) ([]string, bool) {
```
Couldn't we use one of the existing handlebar libs for these kind of things?

blakerouse replied:
I wish I could have used a golang handlebars implementation instead of having to do it myself. But ultimately decided that it was best not to because:

1. Handlebars doesn't support have any notion of `|`. To perform fallbacks or anything like that when it comes to Handlebars you need to do `{{# if ...}` or `{{#unless ...}`.
2. Our implementation needed to know if a variable replace did occur and it came from a dynamic provider that the processors also needed to be attached to the `StrVar` from the `AST`. No Handlebars implementation exposes that information, it is all hidden in there implementation.
3. Our implementation also needed to know if a variable substitution did **not** match. Only then we are able to remove that input because the variable substitution did **not** occur. Again this is hidden in the Handlebars implementations.

ruflin replied:
Makes sense. Your point 1 makes me think if handlebars is perhaps not the right pick as we seem to extend on it and this might be even confusing?

blakerouse replied:
We probably should stop calling it Handlerbars, and just say our variable replacement uses the same syntax as Handlerbars. But it is not Handlebars.

ruflin reviewed (commented):
* Could you add a log message somewhere that this is still experimental? We probably will need it also per provider so we can decide for each provider if it is GA or not.
* What is the current order / overwrite of variables if multiple composable providers are used?

blakerouse commented:
> * Could you add a log message somewhere that this is still experimental? We probably will need it also per provider so we can decide for each provider if it is GA or not.

Yes I will add a log message on startup to document that dynamic inputs are experimental.

> * What is the current order / overwrite of variables if multiple composable providers are used?

What is great about this implementation is you really don't need to worry about order, there is no magic or undefined behavior in ordering.

1. All providers are prefixed with the name of the provider. A providers name is unique, so no overriding will occur.
2. Dynamic providers **never** exist in the same context. So variable replace context will not include say both `docker.*` and `kubernetes.*`.
3. The order of precedence if needed can be determined by the input definition. So a integration package could handle the order of precedence.

Lets say that an input wants to allow a local variable override from the `local` provider, then a variable from the ENV, then fallback to a value from `host` provider.

```
inputs:
  - type: logfile
    streams:
      - paths: /var/log/{{local.path_name|env.PATH_NAME|host.name}}
```

That input defines the order of precedence and Elastic Agent doesn't need to worry about collisions or order of precedence, all because its encoded in the input definition by either the user or the integration package.

Another powerful this is the fallback from one dynamic provider to the next.

```
inputs:
  - type: logfile
    streams:
      - paths: /var/log/{{docker.container.id|kubernetes.container.id}}.log
```

This input will create both inputs for discovered docker containers and kubernetes containers. Because contexts from dynamic providers are never overlapped that means when `docker.container.id` exists `kubernetes.container.id` will never exist and vice versa.

ruflin commented:
I like that there is not "magic" around the order but the user can define it. We had initially that only the dynamic providers are prefixed, but it sounds like now all providers use prefixes, so also `host.name` can only come from the `agent` and not also from environment variables for example? 

The part I stumbled over was `- paths: /var/log/{{docker.container.id|kubernetes.container.id}}.log` as I understand it is possible but I would assume users would not reuse the same input config across multiple dynamic providers. I much prefer your `- paths: /var/log/{{local.path_name|env.PATH_NAME|host.name}}` example.

ruflin commented:
@blakerouse Taking [handlebars discussion](https://github.com/elastic/beats/pull/20839#discussion_r479959981) here so it doesn't get lost when changes are pushed. I just realised that we potentially should not use the same syntax as in the packages for these variables as you initially suggested. We use handlebars in the package so we have if/else/loop support. But this is not something we support on the agent, it is only about variable replacements and applying conditions separately. So using `handlebars` might even give a false impression here. So having a different syntax like `${...|-foo}` might even be helpful.

Note: I know we had a similar conversation somewhere else too but couldn't find it anymore :-(

blakerouse commented:
Moved back to draft while I work on updating to the newest variable syntax.

blakerouse commented:
PR moved back to "Ready for review".

Updated it to use new variable syntax `${env.host|host.name|'default'}`. Prevent go-ucfg from parsing `inputs` from go-ucfg when loading configuration from file or when its coming from Fleet. Improved the variable parsing and replacement logic to be cleaner. Added the ability for full object replacement from variables, instead of previously only allowing strings.

ruflin commented on x-pack/elastic-agent/pkg/agent/application/config.go:
```diff
@@ -26,3 +33,50 @@ func createFleetConfigFromEnroll(accessAPIKey string, kbn *kibana.Config) (*conf
+//
+// This must be used to load the Agent configuration, so that variables defined in the inputs are not
+// parsed by go-ucfg. Variables from the inputs should be parsed by the transpiler.
+func LoadConfigFromFile(path string) (*config.Config, error) {
```
Not official supported but it basically means someone can use `${ELASTIC_USERNAME}` today in the output config because it is ucfg but in case we also support static providers like keystore in the output in the future, this might break as it needs a prefix?

blakerouse replied:
Yes this allows the same go-ucfg that we have always allowed anywhere in the config except in the `inputs`. But if we change that in the future then a prefix would be required to get it from the keystore provider.

ruflin replied:
Lets keep it in mind if we ever there but ok for now.

ruflin commented on x-pack/elastic-agent/pkg/agent/application/emitter.go:
```diff
@@ -30,54 +34,163 @@ type programsDispatcher interface {
 		}
+	}
+
+	e.lock.Lock()
```
I keep stumbling over these lines and the ones on 112 and asking myself if we don't hit any race conditions somewhere or if we could do it in a nicer way.

blakerouse replied:
Being that its internal state to the struct that is called from 2 different go-routines, I don't really have another way of not requiring a lock.

This is also going to be important for the debugging commands to output the current state.

ruflin commented on x-pack/elastic-agent/pkg/agent/application/emitter.go:
```diff
@@ -30,54 +34,163 @@ type programsDispatcher interface {
+		err := e.update()
 		if err != nil {
-			return err
+			e.logger.Errorf("Failed to render configuration with latest context from composable controller: %s", err)
```
Would it be an option to return the error instead of logging it here? This could make testing easier.

blakerouse replied:
It is actually very un-likely that an error will actually occur here, because most errors would come from invalid configuration files, like bad variable syntax or something of that nature. That is why the `Update()` returns the error, because that is caused when the configuration is parsed and will return an error back to the user or back to Fleet.

In the case that a provider changes a value in a variable or a context is removed it will just cause inputs to be add/update/removed. This can also happen at any time, so really logging it here is the only thing we can do.

ruflin commented on x-pack/elastic-agent/pkg/agent/transpiler/vars.go:
```diff
@@ -0,0 +1,199 @@
+	return NewStrValWithProcessors(result+value[lastIndex:], processors), nil
+}
+
+// validBrackets returns true when all starting {{ have a matching ending }}.
```
```suggestion
// validBrackets returns true when all starting ${ have a matching ending }.
```

ruflin commented on x-pack/elastic-agent/pkg/agent/transpiler/vars.go:
```diff
@@ -0,0 +1,199 @@
+	if err != nil {
+		return nil, err
+	}
+	matchIdxs := varsRegex.FindAllSubmatchIndex([]byte(value), -1)
```
This reminds me we need some rules on what chars are allowed in variable names. Should we restrict it?

ruflin replied:
Should we support all of them? Or reduce it to a-z and 2-3 special chars?

ruflin commented on x-pack/elastic-agent/pkg/agent/transpiler/vars_test.go:
```diff
@@ -0,0 +1,242 @@
+		Processors:    processers,
+	}
+
+	//res, err := vars.Replace("${testing.key1}")
```
Leftover?

ruflin commented on x-pack/elastic-agent/pkg/composable/controller.go:
```diff
@@ -8,38 +8,49 @@ import (
-		return nil, errors.New(err, "failed to unpack providers config", errors.TypeConfig)
+		return nil, err
+	}
+	l.Info("EXPERIMENTAL - Dynamic Inputs are currently experimental and should not be used in production")
```
```suggestion
	l.Info("EXPERIMENTAL - Inputs with variables are currently experimental and should not be used in production")
```

ruflin replied:
When exactly is this logged? What kind of config needs the user to have to see this?

blakerouse replied:
This is logged at the start of Elastic Agent when the controller is created. It is always created even without variables in the configuration, so it always logged.

ruflin reviewed (approved):
Tested it locally with the inspect command and seems to work as expected.

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > Either there was a build timeout or someone aborted the build.'}
  


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20839/detail/PR-20839/9//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20839/detail/PR-20839/9//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20839/detail/PR-20839/9//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-20839/detail/PR-20839/9//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #20839 updated]`

* Start Time: 2020-09-03T13:17:02.779+0000

* Duration: 123 min 58 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 1128  |
| Skipped      | 2 |
| Total        | 1130   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Error signal`
    * Description: 
    
    * Duration: 0 min 0 sec
    
    * Start Time: 2020-09-03T15:19:59.311+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-20839/runs/9/steps/958/log)
    
  
  * **Name**: `Cleanup`
    * Description: rm source.tgz
    
    * Duration: 0 min 0 sec
    
    * Start Time: 2020-09-03T15:19:59.356+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-20839/runs/9/steps/961/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-09-03T14:01:33.704Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-20839/src/github.com/elastic/beats/x-pack/elastic-agent
[2020-09-03T14:01:34.497Z] + mage build unitTest
[2020-09-03T14:01:52.353Z] go version go1.14.7 darwin/amd64
[2020-09-03T14:01:52.354Z] GO111MODULE=""
[2020-09-03T14:01:52.354Z] GOARCH="amd64"
[2020-09-03T14:01:52.354Z] GOBIN=""
[2020-09-03T14:01:52.354Z] GOCACHE="/var/lib/jenkins/workspace/Beats_beats_PR-20839/Library/Caches/go-build"
[2020-09-03T14:01:52.354Z] GOENV="/var/lib/jenkins/workspace/Beats_beats_PR-20839/Library/Application Support/go/env"
[2020-09-03T14:01:52.354Z] GOEXE=""
[2020-09-03T14:01:52.354Z] GOFLAGS=""
[2020-09-03T14:01:52.354Z] GOHOSTARCH="amd64"
[2020-09-03T14:01:52.354Z] GOHOSTOS="darwin"
[2020-09-03T14:01:52.354Z] GOINSECURE=""
[2020-09-03T14:01:52.354Z] GONOPROXY=""
[2020-09-03T14:01:52.354Z] GONOSUMDB=""
[2020-09-03T14:01:52.354Z] GOOS="darwin"
[2020-09-03T14:01:52.354Z] GOPATH="/var/lib/jenkins/workspace/Beats_beats_PR-20839"
[2020-09-03T14:01:52.354Z] GOPRIVATE=""
[2020-09-03T14:01:52.354Z] GOPROXY="https://proxy.golang.org,direct"
[2020-09-03T14:01:52.354Z] GOROOT="/var/lib/jenkins/workspace/Beats_beats_PR-20839/.gvm/versions/go1.14.7.darwin.amd64"
[2020-09-03T14:01:52.354Z] GOSUMDB="sum.golang.org"
[2020-09-03T14:01:52.354Z] GOTMPDIR=""
[2020-09-03T14:01:52.354Z] GOTOOLDIR="/var/lib/jenkins/workspace/Beats_beats_PR-20839/.gvm/versions/go1.14.7.darwin.amd64/pkg/tool/darwin_amd64"
[2020-09-03T14:01:52.354Z] GCCGO="gccgo"
[2020-09-03T14:01:52.354Z] AR="ar"
[2020-09-03T14:01:52.354Z] CC="clang"
[2020-09-03T14:01:52.354Z] CXX="clang++"
[2020-09-03T14:01:52.354Z] CGO_ENABLED="1"
[2020-09-03T14:01:52.354Z] GOMOD="/private/var/lib/jenkins/workspace/Beats_beats_PR-20839/src/github.com/elastic/beats/go.mod"
[2020-09-03T14:01:52.354Z] CGO_CFLAGS="-g -O2"
[2020-09-03T14:01:52.354Z] CGO_CPPFLAGS=""
[2020-09-03T14:01:52.354Z] CGO_CXXFLAGS="-g -O2"
[2020-09-03T14:01:52.354Z] CGO_FFLAGS="-g -O2"
[2020-09-03T14:01:52.354Z] CGO_LDFLAGS="-g -O2"
[2020-09-03T14:01:52.354Z] PKG_CONFIG="pkg-config"
[2020-09-03T14:01:52.354Z] GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/pv/j9c15rpx0dv4t3zj2wmjhqsw0000md/T/go-build527479192=/tmp/go-build -gno-record-gcc-switches -fno-common"
[2020-09-03T14:01:52.354Z] >> build: Building elastic-agent
[2020-09-03T14:02:02.799Z] >> go test: Unit Testing
[2020-09-03T15:18:59.292Z] Cancelling nested steps due to timeout
[2020-09-03T15:18:59.360Z] Sending interrupt signal to process
[2020-09-03T15:18:59.888Z] Sending interrupt signal to process
[2020-09-03T15:19:15.063Z] script returned exit code 143
[2020-09-03T15:19:16.055Z] Docker machine "default" does not exist. Use "docker-machine ls" to list machines. Use "docker-machine create" to add a new one.
[2020-09-03T15:19:16.055Z] Docker machine "default" does not exist. Use "docker-machine ls" to list machines. Use "docker-machine create" to add a new one.
[2020-09-03T15:19:16.278Z] Client: Docker Engine - Community
[2020-09-03T15:19:16.278Z]  Version:           19.03.1
[2020-09-03T15:19:16.278Z]  API version:       1.40
[2020-09-03T15:19:16.278Z]  Go version:        go1.12.5
[2020-09-03T15:19:16.278Z]  Git commit:        74b1e89
[2020-09-03T15:19:16.278Z]  Built:             Thu Jul 25 21:18:17 2019
[2020-09-03T15:19:16.278Z]  OS/Arch:           darwin/amd64
[2020-09-03T15:19:16.278Z]  Experimental:      false
[2020-09-03T15:19:16.278Z] Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
[2020-09-03T15:19:16.278Z] It requires Docker daemon to be installed and running
[2020-09-03T15:19:17.147Z] + python .ci/scripts/pre_archive_test.py
[2020-09-03T15:19:17.628Z] Copy ./x-pack/elastic-agent/build into build/x-pack/elastic-agent/build
[2020-09-03T15:19:17.804Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-20839/src/github.com/elastic/beats/build
[2020-09-03T15:19:17.895Z] Recording test results
[2020-09-03T15:19:18.055Z] None of the test reports contained any result
[2020-09-03T15:19:18.155Z] Stashed 0 file(s)
[2020-09-03T15:19:18.221Z] Archiving artifacts
[2020-09-03T15:19:19.211Z] + python .ci/scripts/search_system_tests.py
[2020-09-03T15:19:19.447Z] [INFO] system-tests=''. If no empty then let's create a tarball
[2020-09-03T15:19:20.778Z] Post stage
[2020-09-03T15:19:20.787Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-20839/src/github.com/elastic/beats
[2020-09-03T15:19:21.701Z] Starting "default"...
[2020-09-03T15:19:21.701Z] Machine "default" is already running.
[2020-09-03T15:19:23.292Z] Error checking TLS connection: Error checking and/or regenerating the certs: There was an error validating certificates for host "192.168.99.101:2376": dial tcp 192.168.99.101:2376: connect: connection refused
[2020-09-03T15:19:23.292Z] You can attempt to regenerate them using 'docker-machine regenerate-certs [name]'.
[2020-09-03T15:19:23.292Z] Be advised that this will trigger a Docker daemon restart which might stop running containers.
[2020-09-03T15:19:23.292Z] 
[2020-09-03T15:19:23.292Z] Client: Docker Engine - Community
[2020-09-03T15:19:23.292Z]  Version:           19.03.1
[2020-09-03T15:19:23.292Z]  API version:       1.40
[2020-09-03T15:19:23.292Z]  Go version:        go1.12.5
[2020-09-03T15:19:23.292Z]  Git commit:        74b1e89
[2020-09-03T15:19:23.292Z]  Built:             Thu Jul 25 21:18:17 2019
[2020-09-03T15:19:23.292Z]  OS/Arch:           darwin/amd64
[2020-09-03T15:19:23.292Z]  Experimental:      false
[2020-09-03T15:19:23.292Z] Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
[2020-09-03T15:19:23.292Z] It requires Docker daemon to be installed and running
[2020-09-03T15:19:35.082Z] Failed in branch Elastic Agent Mac OS X
[2020-09-03T15:19:43.999Z] [INFO] unstashV2: JOB_GCS_BUCKET is set. bucket param got precedency instead.
[2020-09-03T15:19:44.021Z] [INFO] unstashV2: JOB_GCS_CREDENTIALS is set. credentialsId param got precedency instead.
[2020-09-03T15:19:44.085Z] [Google Cloud Storage Plugin] Found 1 files to download from pattern: gs://beats-ci-temp/Beats/beats/PR-20839-9/source/source.tgz
[2020-09-03T15:19:44.102Z] [Google Cloud Storage Plugin] Downloading: Beats/beats/PR-20839-9/source/source.tgz to local path: /var/lib/jenkins/workspace/Beats_beats_PR-20839/source.tgz
[2020-09-03T15:19:53.261Z] + tar --version
[2020-09-03T15:19:53.557Z] + tar -xpf source.tgz
[2020-09-03T15:19:59.293Z] Body did not finish within grace period; terminating with extreme prejudice
[2020-09-03T15:19:59.309Z] [INFO] source.tgz was not extracted : null
[2020-09-03T15:19:59.648Z] + rm source.tgz
[2020-09-03T15:19:59.657Z] ERROR: runbld post build action failed.
[2020-09-03T15:19:59.657Z] ERROR: untar: step failed with error null
[2020-09-03T15:19:59.860Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-20839
[2020-09-03T15:19:59.968Z] [INFO] getVaultSecret: Getting secrets
[2020-09-03T15:20:00.040Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-09-03T15:20:00.915Z] + chmod 755 generate-build-data.sh
[2020-09-03T15:20:00.915Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20839/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20839/runs/9 ABORTED 7377873
[2020-09-03T15:20:00.915Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20839/runs/9/steps/?limit=10000 -o steps-info.json
[2020-09-03T15:20:01.826Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-20839/runs/9/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->
